public/manage/github.php

```diff
@@ -70,28 +79,75 @@ function github_require_valid_user()
       'header'  => 'Content-type: application/x-www-form-urlencoded',
       'content' => $data_encoded,
     ];
-    $ctxt = stream_context_create(['http' => $opts]);
-    $s = @file_get_contents('https://github.com/login/oauth/access_token', false, $ctxt);
+    $context = stream_context_create(['http' => $opts]);
+    $s = @file_get_contents('https://github.com/login/oauth/access_token', false, $context);
     if (!$s) {
-      die('Failed while checking with GitHub,either you are trying to hack us or our configuration is wrong (GITHUB_CLIENT_SECRET outdated?)');
+      head("github administration");
+      warn("Failed while checking with GitHub,either you are trying to hack us or our configuration is wrong (GITHUB_CLIENT_SECRET outdated?)");
+      foot();
+      exit;
     }
     $gh = [];
     parse_str($s, $gh);
     if (empty($gh['access_token'])) {
-      die("GitHub responded but didn't send an access_token");
+      head("github administration");
+      warn("GitHub responded but didn't send an access_token");
+      foot();
+      exit;
     }
 
-    $user = github_current_user($gh['access_token']);
+    $user = github_current_user($gh["access_token"]);
+
+    $opts = ['user_agent' => GITHUB_USER_AGENT, 'header' => 'Authorization: token '. urlencode($gh['access_token'])."\r\n"];
+    $context = stream_context_create(['http' => $opts]);
+    file_get_contents('https://api.github.com/orgs/php/members/'.urlencode($user->login), false, $context);
+    $statusLine = $http_response_header[0];
+    preg_match('{HTTP\/\S*\s(\d{3})}', $statusLine, $match);
+    $status = $match[1];
+    $is_member = (int)$status === 204;
 
-    $endpoint = '/teams/'.urlencode((string)GITHUB_PHP_OWNER_TEAM_ID).'/members/'.urlencode($user->login);
     $opts = ['user_agent' => GITHUB_USER_AGENT];
-    $ctxt = stream_context_create(['http' => $opts]);
-    $is_member = file_get_contents('https://api.github.com'.$endpoint.'?access_token='.urlencode($gh['access_token']), false, $ctxt);
+    $context = stream_context_create(['http' => $opts]);
+    $is_admin = file_get_contents('https://api.github.com/teams/'.urlencode((string)GITHUB_PHP_OWNER_TEAM_ID).'/members/'.urlencode($user->login).'?access_token='.urlencode($gh['access_token']), false, $context);
 
     if ($is_member === false) {
       head("github administration");
       echo '<h1>You (Authenticated GitHub user: '.htmlentities($user->login). ') are no member of the php organization on github.</h1>'.
-        '<p>Please contact an existing member if you see need.</p>';
+          '<p>Please contact an existing member if you see need.</p>';
+      foot();
+      exit;
+    }
```

Comment:
@nikic,

I was thinking of a scenario like this:

we do not ask the user for additional scopes, just a name is enough.
Add a user with administrator rights to the PHP organization (e.g. @php-pulls or new).

When authorizing through GitHub, we link accounts and also using the admin user token, we check if there is a current user in the organization. If not yet, send an invite.

Is this workflow valid from an organizational security perspective?
Or is it preferable to leave control manually and through this script just link accounts?

Reply:
In the end, it will still be necessary to add the user to the correct team or grant repository permissions. An invite is automatically sent as part of doing that. As such, I don't think it's necessary to automate the sending of the invite here.